File: sdk/humanlayer/ai/callback_handler.py
"""LangChain callback handler for HumanLayer event capture."""
import uuid
import logging
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Union

from langchain_core.callbacks.base import BaseCallbackHandler

from humanlayer.ai.session import get_current_session

logger = logging.getLogger("humanlayer.callback")


class HumanLayerCallbackHandler(BaseCallbackHandler):
    """Captures all LangChain/LangGraph lifecycle events with standardized format."""

    # Always run regardless of other handlers
    raise_error = False

    def __init__(self, session=None):
        super().__init__()
        self._session = session
        self._run_id = str(uuid.uuid4())

    @property
    def session(self):
        return self._session or get_current_session()

    # ── LLM Events ──────────────────────────────────────────────────────────

    def on_chat_model_start(self, serialized, messages, *, run_id, parent_run_id=None, tags=None, metadata=None, **kwargs):
        name = self._extract_name(serialized)
        model = self._extract_model(serialized)
        flat_msgs = []
        for batch in messages:
            for m in batch:
                flat_msgs.append(getattr(m, "content", str(m)))
        self._log("llm_start", "llm", {
            "run_id": str(run_id), "parent_run_id": str(parent_run_id) if parent_run_id else None,
            "name": name, "model": model,
            "messages": flat_msgs[:3],
        })

    def on_llm_start(self, serialized, prompts, *, run_id, parent_run_id=None, **kwargs):
        self._log("llm_start", "llm", {
            "run_id": str(run_id), "parent_run_id": str(parent_run_id) if parent_run_id else None,
            "name": self._extract_name(serialized),
            "model": self._extract_model(serialized),
            "prompts": prompts[:2],
        })

    def on_llm_end(self, response, *, run_id, parent_run_id=None, **kwargs):
        output = None
        usage = {}
        try:
            gen = response.generations
            if gen and gen[0]:
                output = getattr(gen[0][0], "text", None) or str(gen[0][0])
        except Exception:
            pass
        try:
            usage = response.llm_output.get("token_usage", {}) if response.llm_output else {}
        except Exception:
            pass
        self._log("llm_end", "llm", {
            "run_id": str(run_id),
            "output": output,
            "token_usage": usage,
        })

    def on_llm_error(self, error, *, run_id, **kwargs):
        self._log("llm_error", "llm", {"run_id": str(run_id), "error": str(error)})

    # ── Tool Events ──────────────────────────────────────────────────────────

    def on_tool_start(self, serialized, input_str, *, run_id, parent_run_id=None, **kwargs):
        self._log("tool_start", "tool", {
            "run_id": str(run_id), "parent_run_id": str(parent_run_id) if parent_run_id else None,
            "name": self._extract_name(serialized),
            "inputs": input_str,
        })

    def on_tool_end(self, output, *, run_id, **kwargs):
        self._log("tool_end", "tool", {
            "run_id": str(run_id),
            "output": self._serialize(output),
        })

    def on_tool_error(self, error, *, run_id, **kwargs):
        self._log("tool_error", "tool", {"run_id": str(run_id), "error": str(error)})

    # ── Chain Events ─────────────────────────────────────────────────────────

    def on_chain_start(self, serialized, inputs, *, run_id, parent_run_id=None, **kwargs):
        name = self._extract_name(serialized)
        self._log("chain_start", "chain", {
            "run_id": str(run_id), "parent_run_id": str(parent_run_id) if parent_run_id else None,
            "name": name, "inputs": self._serialize(inputs),
        })

    def on_chain_end(self, outputs, *, run_id, **kwargs):
        self._log("chain_end", "chain", {
            "run_id": str(run_id),
            "outputs": self._serialize(outputs),
        })

    def on_chain_error(self, error, *, run_id, **kwargs):
        self._log("chain_error", "chain", {"run_id": str(run_id), "error": str(error)})

    # ── Agent Events ─────────────────────────────────────────────────────────

    def on_agent_action(self, action, *, run_id, **kwargs):
        self._log("agent_action", "agent", {
            "run_id": str(run_id),
            "tool": getattr(action, "tool", "unknown"),
            "tool_input": self._serialize(getattr(action, "tool_input", {})),
            "log": getattr(action, "log", ""),
        })

    def on_agent_finish(self, finish, *, run_id, **kwargs):
        rv = getattr(finish, "return_values", {})
        self._log("agent_finish", "agent", {
            "run_id": str(run_id),
            "output": self._serialize(rv.get("output", rv)),
            "log": getattr(finish, "log", ""),
        })

    # ── Internals ────────────────────────────────────────────────────────────

    def _log(self, event_type: str, component: str, raw_data: dict) -> None:
        session = self.session
        if not session:
            return

        data = self._standardize(raw_data)
        if self._should_filter(data, event_type):
            return

        event = {
            "event_id": str(uuid.uuid4()),
            "run_id": self._run_id,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "event_type": event_type,
            "component": component,
            "data": data,
        }
        session.log_event(event)

    def _standardize(self, raw: dict) -> dict:
        out: dict = {}

        for k in ("run_id", "parent_run_id", "name", "model", "tool"):
            if k in raw and raw[k] is not None:
                out[k] = raw[k]

        # Standardise input
        inp = None
        for key in ("inputs", "prompts", "messages", "tool_input"):
            if key in raw:
                inp = raw[key]
                break
        if inp:
            cleaned = self._clean(inp)
            if cleaned:
                out["input"] = cleaned

        # Standardise output
        outp = None
        for key in ("output", "outputs", "response"):
            if key in raw:
                outp = raw[key]
                break
        if outp:
            cleaned = self._clean(outp)
            if cleaned:
                out["output"] = cleaned

        # Metadata bucket
        meta = {}
        if raw.get("log"):
            meta["agent_log"] = raw["log"][:300]
        if raw.get("token_usage"):
            meta["token_usage"] = raw["token_usage"]
        if raw.get("error"):
            meta["error"] = str(raw["error"])[:300]
        if meta:
            out["metadata"] = meta

        return out

    def _should_filter(self, data: dict, event_type: str) -> bool:
        if event_type in ("agent_action", "agent_finish") or "error" in event_type:
            return False
        if event_type in ("tool_start", "tool_end"):
            return False
        if event_type in ("llm_start", "llm_end"):
            return not ("input" in data or "output" in data or "model" in data)
        if event_type in ("chain_start", "chain_end"):
            name = data.get("name", "")
            keep = {"AgentExecutor", "RunnableSequence", "PromptTemplate"}
            return name not in keep and "input" not in data
        return False

    def _clean(self, value) -> Any:
        s = self._serialize(value)
        if not s:
            return None
        if isinstance(s, str) and len(s) > 1000:
            return s[:1000] + "…"
        return s

    def _serialize(self, value, depth=0) -> Any:
        if depth > 3 or value is None:
            return None
        if isinstance(value, (str, int, float, bool)):
            return value
        if isinstance(value, dict):
            return {k: self._serialize(v, depth + 1) for k, v in list(value.items())[:20]}
        if isinstance(value, (list, tuple)):
            return [self._serialize(v, depth + 1) for v in list(value)[:10]]
        if hasattr(value, "content"):
            return getattr(value, "content", "")
        if hasattr(value, "dict"):
            try:
                return self._serialize(value.dict(), depth + 1)
            except Exception:
                pass
        return str(value)[:500]

    def _extract_name(self, serialized) -> str:
        if not serialized:
            return "unknown"
        return (
            serialized.get("name")
            or serialized.get("id", ["unknown"])[-1]
            or "unknown"
        )

    def _extract_model(self, serialized) -> Optional[str]:
        if not serialized:
            return None
        kw = serialized.get("kwargs", {})
        return kw.get("model_name") or kw.get("model") or serialized.get("name")

File: sdk-js/src/callbackHandler.ts
import { BaseCallbackHandler } from "@langchain/core/callbacks/base";
import type { LLMResult } from "@langchain/core/outputs";
import type { ChainValues } from "@langchain/core/utils/types";
import { v4 as uuidv4 } from "uuid";
import { getCurrentSession } from "./session";

// Noise filter: these chain names produce redundant events
const FILTERED_CHAIN_NAMES = new Set([
  "RunnableSequence",
  "RunnableParallel",
  "RunnableLambda",
  "RunnablePassthrough",
  "RunnableMap",
]);

function truncate(value: unknown, maxLen = 1000): unknown {
  if (typeof value === "string") {
    return value.length > maxLen ? value.slice(0, maxLen) + "...[truncated]" : value;
  }
  return value;
}

function cleanData(obj: unknown, depth = 0): unknown {
  if (depth > 3) return "[nested]";
  if (obj === null || obj === undefined) return obj;
  if (typeof obj === "string") return truncate(obj);
  if (typeof obj === "number" || typeof obj === "boolean") return obj;
  if (Array.isArray(obj)) {
    return obj.slice(0, 20).map((item) => cleanData(item, depth + 1));
  }
  if (typeof obj === "object") {
    const result: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(obj as Record<string, unknown>)) {
      result[k] = cleanData(v, depth + 1);
    }
    return result;
  }
  return String(obj);
}

function extractModelName(serialized: Record<string, unknown>): string | undefined {
  const id = serialized?.id;
  const lastId = Array.isArray(id) ? (id as string[]).slice(-1)[0] : undefined;
  return (
    (serialized?.model_name as string) ??
    (serialized?.model as string) ??
    lastId ??
    undefined
  );
}

function extractMessages(messages: unknown): unknown {
  if (!Array.isArray(messages)) return messages;
  return messages.map((msgGroup) => {
    if (Array.isArray(msgGroup)) {
      return msgGroup.map((m) => {
        if (m && typeof m === "object") {
          const msg = m as Record<string, unknown>;
          return {
            role: Array.isArray(msg.lc_id) ? (msg.lc_id as string[]).slice(-1)[0]?.toLowerCase().replace("message", "") ?? "unknown" : "unknown",
            content: truncate(msg.content ?? msg.text ?? ""),
          };
        }
        return m;
      });
    }
    return msgGroup;
  });
}

export class HumanLayerCallbackHandler extends BaseCallbackHandler {
  name = "HumanLayerCallbackHandler";
  private runId: string;

  constructor() {
    super();
    this.runId = uuidv4();
  }

  private _emit(
    eventType: string,
    component: string,
    data: Record<string, unknown>
  ) {
    const session = getCurrentSession();
    const event = {
      event_id: uuidv4(),
      run_id: this.runId,
      timestamp: new Date().toISOString(),
      event_type: eventType,
      component,
      data,
    };
    if (session) {
      session.logEvent(event);
    }
  }

  override async handleChatModelStart(
    llm: { id?: string[]; kwargs?: Record<string, unknown> },
    messages: unknown[][],
    runId: string,
    parentRunId?: string
  ): Promise<void> {
    this._emit("llm_start", "llm", {
      run_id: runId,
      parent_run_id: parentRunId,
      name: llm.id?.slice(-1)[0] ?? "unknown_llm",
      model: extractModelName(llm.kwargs ?? {}),
      input: cleanData(extractMessages(messages)),
      output: null,
      metadata: {},
    });
  }

  override async handleLLMStart(
    llm: { id?: string[]; kwargs?: Record<string, unknown> },
    prompts: string[],
    runId: string,
    parentRunId?: string
  ): Promise<void> {
    this._emit("llm_start", "llm", {
      run_id: runId,
      parent_run_id: parentRunId,
      name: llm.id?.slice(-1)[0] ?? "unknown_llm",
      model: extractModelName(llm.kwargs ?? {}),
      input: prompts.map((p) => truncate(p)),
      output: null,
      metadata: {},
    });
  }

  override async handleLLMEnd(
    output: LLMResult,
    runId: string,
    parentRunId?: string
  ): Promise<void> {
    const usage = output.llmOutput?.tokenUsage ?? output.llmOutput?.usage ?? {};
    const gen = output.generations?.[0]?.[0];
    const text = gen?.text ?? (gen as unknown as Record<string, unknown>)?.message;

    this._emit("llm_end", "llm", {
      run_id: runId,
      parent_run_id: parentRunId,
      name: "llm",
      output: cleanData(text),
      metadata: {
        token_usage: {
          total_tokens:
            (usage as Record<string, unknown>).totalTokens ??
            (usage as Record<string, unknown>).total_tokens ??
            0,
          prompt_tokens:
            (usage as Record<string, unknown>).promptTokens ??
            (usage as Record<string, unknown>).prompt_tokens ??
            0,
          completion_tokens:
            (usage as Record<string, unknown>).completionTokens ??
            (usage as Record<string, unknown>).completion_tokens ??
            0,
        },
      },
    });
  }

  override async handleLLMError(
    err: Error,
    runId: string,
    parentRunId?: string
  ): Promise<void> {
    this._emit("llm_error", "llm", {
      run_id: runId,
      parent_run_id: parentRunId,
      name: "llm",
      metadata: { error: err.message },
    });
  }

  override async handleToolStart(
    tool: { id?: string[]; name?: string },
    input: string,
    runId: string,
    parentRunId?: string
  ): Promise<void> {
    const toolName =
      tool.name ?? tool.id?.slice(-1)[0] ?? "unknown_tool";
    this._emit("tool_start", "tool", {
      run_id: runId,
      parent_run_id: parentRunId,
      name: toolName,
      input: cleanData(
        (() => {
          try {
            return JSON.parse(input);
          } catch {
            return input;
          }
        })()
      ),
      output: null,
      metadata: {},
    });
  }

  override async handleToolEnd(
    output: string,
    runId: string,
    parentRunId?: string
  ): Promise<void> {
    this._emit("tool_end", "tool", {
      run_id: runId,
      parent_run_id: parentRunId,
      name: "tool",
      output: cleanData(
        (() => {
          try {
            return JSON.parse(output);
          } catch {
            return output;
          }
        })()
      ),
      metadata: {},
    });
  }

  override async handleToolError(
    err: Error,
    runId: string,
    parentRunId?: string
  ): Promise<void> {
    this._emit("tool_error", "tool", {
      run_id: runId,
      parent_run_id: parentRunId,
      name: "tool",
      metadata: { error: err.message },
    });
  }

  override async handleChainStart(
    chain: { id?: string[] },
    inputs: ChainValues,
    runId: string,
    parentRunId?: string
  ): Promise<void> {
    const chainName = chain.id?.slice(-1)[0] ?? "chain";
    if (FILTERED_CHAIN_NAMES.has(chainName)) return;
    this._emit("chain_start", "chain", {
      run_id: runId,
      parent_run_id: parentRunId,
      name: chainName,
      input: cleanData(inputs),
      output: null,
      metadata: {},
    });
  }

  override async handleChainEnd(
    outputs: ChainValues,
    runId: string,
    parentRunId?: string
  ): Promise<void> {
    this._emit("chain_end", "chain", {
      run_id: runId,
      parent_run_id: parentRunId,
      name: "chain",
      output: cleanData(outputs),
      metadata: {},
    });
  }

  override async handleChainError(
    err: Error,
    runId: string,
    parentRunId?: string
  ): Promise<void> {
    this._emit("chain_error", "chain", {
      run_id: runId,
      parent_run_id: parentRunId,
      name: "chain",
      metadata: { error: err.message },
    });
  }

  override async handleAgentAction(
    action: { tool: string; toolInput: unknown; log: string },
    runId: string,
    parentRunId?: string
  ): Promise<void> {
    this._emit("agent_action", "agent", {
      run_id: runId,
      parent_run_id: parentRunId,
      name: action.tool,
      input: cleanData(action.toolInput),
      metadata: { agent_log: truncate(action.log) as string },
    });
  }

  override async handleAgentEnd(
    action: { returnValues: unknown; log: string },
    runId: string,
    parentRunId?: string
  ): Promise<void> {
    this._emit("agent_finish", "agent", {
      run_id: runId,
      parent_run_id: parentRunId,
      name: "agent",
      output: cleanData(action.returnValues),
      metadata: { agent_log: truncate(action.log) as string },
    });
  }
}
